runtime/observability/logging-slf4j2/jvm/src/aws/smithy/kotlin/runtime/telemetry/logging/slf4j/Slf4jLoggerProvider.kt

```diff
@@ -5,73 +5,24 @@
 
 package aws.smithy.kotlin.runtime.telemetry.logging.slf4j
 
-import aws.smithy.kotlin.runtime.telemetry.context.Context
 import aws.smithy.kotlin.runtime.telemetry.logging.*
 import org.slf4j.LoggerFactory
-import org.slf4j.event.Level
-import org.slf4j.spi.LoggingEventBuilder
 
 /**
- * SLF4J 2 based logger provider
+ * SLF4J based logger provider
  */
 public object Slf4jLoggerProvider : LoggerProvider {
-    override fun getOrCreateLogger(name: String): Logger {
-        val sl4fjLogger = LoggerFactory.getLogger(name)
-        return Slf4JLoggerAdapter(sl4fjLogger)
-    }
-}
-
-private class Slf4JLoggerAdapter(private val delegate: org.slf4j.Logger) : Logger {
-    private fun logWith(t: Throwable?, msg: () -> String, builder: LoggingEventBuilder) =
-        builder.setMessage(msg)
-            .apply {
-                if (t != null) {
-                    setCause(t)
-                }
-            }.log()
-    override fun trace(t: Throwable?, msg: () -> String) = logWith(t, msg, delegate.atTrace())
-    override fun debug(t: Throwable?, msg: () -> String) = logWith(t, msg, delegate.atDebug())
-    override fun info(t: Throwable?, msg: () -> String) = logWith(t, msg, delegate.atInfo())
-    override fun warn(t: Throwable?, msg: () -> String) = logWith(t, msg, delegate.atWarn())
-    override fun error(t: Throwable?, msg: () -> String) = logWith(t, msg, delegate.atError())
-    override fun isEnabledFor(level: LogLevel): Boolean =
-        delegate.isEnabledForLevel(level.slf4jLevel)
-
-    override fun atLevel(level: LogLevel): LogRecordBuilder =
-        Slf4jLogRecordBuilderAdapter(delegate.atLevel(level.slf4jLevel))
-}
-
-private class Slf4jLogRecordBuilderAdapter(
-    private val delegate: LoggingEventBuilder,
-) : LogRecordBuilder {
-    override fun setCause(ex: Throwable) {
-        delegate.setCause(ex)
-    }
 
-    override fun setMessage(message: String) {
-        delegate.setMessage(message)
+    private val slf4jLoggerAdapter: (org.slf4j.Logger) -> Logger = try {
+        Class.forName("org.slf4j.spi.LoggingEventBuilder")
+        ::Slf4j2xLoggerAdapter
+    } catch (ex: ClassNotFoundException) {
+        LoggerFactory.getLogger(Slf4jLoggerProvider::class.java).warn("falling back to SLF4J 1.x compatible binding")
+        ::Slf4j1xLoggerAdapter
     }
```

Comment:
**Style**: May be clearer as a function.

Reply:
Not opposed but can you clarify what you are referring to? What would you prefer to see here?

Reply:
A class-level `val` that returns a function is more complex to read than simply a `fun`:

```kotlin
private fun slf4jLoggerAdapter(delegate: org.slf4j.Logger) = try {
    Class.forName("org.slf4j.spi.LoggingEventBuilder")
    Slf4j2xLoggerAdapter(delegate)
} catch (_: ClassNotFoundException) {
    LoggerFactory.getLogger(Slf4jLoggerProvider::class.java).warn("falling back to SLF4J 1.x compatible binding")
    Slf4j1xLoggerAdapter(delegate)
}
```

Reply:
I was trying to avoid invoking the class loader on every logger instantiation.

Reply:
I could refactor it though to something like:


```kotlin
private val slf4j2Found = try {
    Class.forName("org.slf4j.spi.LoggingEventBuilder")
    true
} catch(_: ClassNotFoundException) {
     LoggerFactory.getLogger(Slf4jLoggerProvider::class.java).warn("falling back to SLF4J 1.x compatible binding")
    false  
}

   override fun getOrCreateLogger(name: String): Logger {
        val sl4fjLogger = LoggerFactory.getLogger(name)
        return if (slf4j2Found) { 
             Slf4j2xLoggerAdapter(slf4jLogger)
        } else {
             Slf4j1xLoggerAdapter(slf4jLogger)
        }
    }

```

Reply:
Oh right. I didn't think about caching the class loader invocation. OK, your original code works better then.